core/trino-main/src/main/java/io/trino/sql/planner/iterative/IterativeOptimizer.java

```diff
@@ -151,19 +157,29 @@ private boolean exploreNode(int group, Context context)
             Iterator<Rule<?>> possiblyMatchingRules = ruleIndex.getCandidates(node).iterator();
             while (possiblyMatchingRules.hasNext()) {
                 Rule<?> rule = possiblyMatchingRules.next();
-
-                if (!rule.isEnabled(context.session)) {
-                    continue;
+                long timeStart = nanoTime();
+                long timeEnd;
+                boolean invoked = false;
+                boolean applied = false;
+
+                if (rule.isEnabled(context.session)) {
+                    invoked = true;
+                    Rule.Result result = transform(node, rule, context);
+                    timeEnd = nanoTime();
+
+                    if (result.getTransformedPlan().isPresent()) {
+                        node = context.memo.replace(group, result.getTransformedPlan().get(), rule.getClass().getName());
+
+                        applied = true;
+                        done = false;
+                        progress = true;
+                    }
                 }
-
-                Rule.Result result = transform(node, rule, context);
-
-                if (result.getTransformedPlan().isPresent()) {
-                    node = context.memo.replace(group, result.getTransformedPlan().get(), rule.getClass().getName());
-
-                    done = false;
-                    progress = true;
+                else {
+                    timeEnd = nanoTime();
```

Comment:
The reason for structuring the code this way is that i didn't want to assume `isEnabled` takes no time.

Reply:
add a comment please

Reply:
missed this one